contracts/strategies/traderjoe/TraderJoe.sol

```diff
@@ -84,62 +171,76 @@ contract TraderJoe is UUPSUpgradeable, StrategyOwnablePausableBaseUpgradeable {
         uint256 swapAmount = amount / 2;
         address[] memory path = new address[](2);
         path[0] = address(depositToken);
         path[1] = address(strategyStorage.pairDepositToken);
 
-        uint256 pairDepositTokenDesired = SwapServiceLib
+        uint256 pairDepositTokenAmount = SwapServiceLib
             .swapExactTokensForTokens(swapService, swapAmount, 0, path);
-        uint256 depositTokenDesired = amount - swapAmount;
+        uint256 depositTokenAmount = amount - swapAmount;
 
-        // providing liquidity to Trader Joe
-        strategyStorage.pairDepositToken.approve(
-            address(strategyStorage.router),
-            pairDepositTokenDesired
-        );
-        depositToken.approve(
-            address(strategyStorage.router),
-            depositTokenDesired
-        );
-        (, , uint256 lpBalance) = strategyStorage.router.addLiquidity(
-            address(strategyStorage.pairDepositToken),
-            address(depositToken),
-            pairDepositTokenDesired,
-            depositTokenDesired,
-            0,
-            0,
-            address(this),
-            // solhint-disable-next-line not-rely-on-time
-            block.timestamp
-        );
+        uint256 amountX;
+        uint256 amountY;
+        if (strategyStorage.tokenX == depositToken) {
+            amountX = depositTokenAmount;
+            amountY = pairDepositTokenAmount;
+        } else {
+            amountX = pairDepositTokenAmount;
+            amountY = depositTokenAmount;
+        }
 
-        // swapping back extra pairDepositToken to depositToken
-        uint256 extraPairDepositTokenAmount = strategyStorage
-            .pairDepositToken
-            .balanceOf(address(this)) - pairDepositTokenBalanceBeforeSwap;
-        if (extraPairDepositTokenAmount != 0) {
-            path[0] = address(strategyStorage.pairDepositToken);
-            path[1] = address(depositToken);
-            SwapServiceLib.swapExactTokensForTokens(
-                swapService,
-                extraPairDepositTokenAmount,
+        // Prepare params.
+        uint256 binsAmount = strategyStorage.binIds.length;
+        uint256[] memory distributionX = new uint256[](binsAmount);
+        uint256[] memory distributionY = new uint256[](binsAmount);
+
+        for (uint256 i; i < binsAmount; ++i) {
+            // Bin allocation has precision of 1e3 and TraderJoe V2 has 1e18.
+            distributionX[i] = strategyStorage.binAllocationsX[i] * 1e15;
+            distributionY[i] = strategyStorage.binAllocationsY[i] * 1e15;
+        }
+
+        int256[] memory deltaIds = new int256[](binsAmount);
+        (, , uint256 activeId) = strategyStorage.lbPair.getReservesAndId();
+
+        for (uint256 i; i < binsAmount; ++i) {
+            deltaIds[i] = int256(strategyStorage.binIds[i]) - int256(activeId);
+        }
+
+        // Deposit.
+        ITraderJoeLBRouter.LiquidityParameters memory liquidityParameters = ITraderJoeLBRouter
```

Comment:
So let's say the active bin now is 100, and we provide liquidity, that means now our contract provide liquidity for 50 bins. Some time after that the active bin is 200 and we provide liquidity again. That means our contract now has 100 different bins where we provide liquidity